main.c

```diff
@@ -275,11 +275,23 @@ static void elf_boot(const struct boot_params params)
         const uint8_t *current_elf_segment_in_memory =
             elf_addr_in_memory + current_program_header->offset;
 
-        printf(
-            "Loading ELF segment (type 0x%x, offset 0x%x, vaddr 0x%x, paddr 0x%x, filesize 0x%x, memsize 0x%x)\n",
-            current_program_header->type, current_program_header->offset,
-            current_program_header->vaddr, current_program_header->paddr,
-            current_program_header->filesize, current_program_header->memsize);
+        if (current_program_header->type != 1 /* LOAD segment */) {
+            printf(
+                "Skipping Non-LOAD ELF segment (type 0x%x, offset 0x%x, vaddr 0x%x, paddr 0x%x, filesize 0x%x, memsize 0x%x)\n",
+                current_program_header->type, current_program_header->offset,
+                current_program_header->vaddr, current_program_header->paddr,
+                current_program_header->filesize, current_program_header->memsize);
+            continue;
+        } else {
+            printf(
+                "Loading ELF segment (type 0x%x, offset 0x%x, vaddr 0x%x, paddr 0x%x, filesize 0x%x, memsize 0x%x)\n",
+                current_program_header->type, current_program_header->offset,
+                current_program_header->vaddr, current_program_header->paddr,
+                current_program_header->filesize, current_program_header->memsize);
+        }
+
+        die_on(current_program_header->filesize > current_program_header->memsize,
+               "filesize > memsize. The ELF segment looks broken.\n");
```

Comment:
I'm fine with adding this safety check. But have you seen this on a legitimate ELF object or was that a handcrafted experiment?

Reply:
```
Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  LOAD           0x001000 0x0000000000800000 0x0000000000800000 0x00afad 0x00afad   E 0x1000
  LOAD           0x00c000 0x000000000080b000 0x000000000080b000 0x000c08 0x000c08 R   0x1000
  LOAD           0x00d000 0x000000000080c000 0x000000000080c000 0x0070d0 0x1090f0 RW  0x1000
  NOTE           0x0140d0 0x00000000009150f0 0x00000000009150f0 0x000014 0x000000 R   0x4
```

However, we filter that out by checking the segment type already.